xrootd/client/client.go

```diff
@@ -0,0 +1,141 @@
+// Copyright 2018 The go-hep Authors.  All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package client // import "go-hep.org/x/hep/xrootd/client"
+
+import (
+	"context"
+	"encoding/binary"
+	"io"
+	"net"
+
+	"go-hep.org/x/hep/xrootd/encoder"
+	"go-hep.org/x/hep/xrootd/mux"
+	"go-hep.org/x/hep/xrootd/protocol"
+)
+
+// A Client to xrootd server which allows to send requests and receive responses.
+// Concurrent requests are supported.
+// Zero value is invalid, Client should be instantiated using NewClient.
+type Client struct {
+	conn            net.Conn
+	mux             *mux.Mux
+	protocolVersion int32
+}
+
+const responseHeaderSize = 2 + 2 + 4
+
+type responseHeader struct {
+	StreamID   protocol.StreamID
+	Status     uint16
+	DataLength int32
+}
+
+// NewClient creates a client to xrootd server at address.
+// ctx defines a lifetime of the client, once it is done response handling is stopped.
+func NewClient(ctx context.Context, address string) (*Client, error) {
+	conn, err := net.Dial("tcp", address)
+	if err != nil {
+		return nil, err
+	}
+
+	client := &Client{conn, mux.New(), 0}
+
+	go client.consume(ctx)
+
+	if err := client.handshake(ctx); err != nil {
+		return nil, err
+	}
+
+	return client, nil
+}
+
+func (client *Client) consume(ctx context.Context) {
+	var header = &responseHeader{}
+	var headerBytes = make([]byte, responseHeaderSize)
+
+	for {
+		if _, err := io.ReadFull(client.conn, headerBytes); err != nil {
+			// TODO: handle EOF by redirection as specified at http://xrootd.org/doc/dev45/XRdv310.pdf, page 11
+		}
+
+		if err := encoder.Unmarshal(headerBytes, header); err != nil {
+			// TODO: should redirect in case if is not possible to decode a header as well?
+		}
+
+		resp := mux.ServerResponse{make([]byte, header.DataLength), nil}
+		if _, err := io.ReadFull(client.conn, resp.Data); err != nil {
+			resp.Err = err
+		} else if header.Status != protocol.Ok {
+			resp.Err = extractError(header, resp.Data)
+		}
+
+		if err := client.mux.SendData(header.StreamID, resp); err != nil {
+			// TODO: should we just ignore responses to unclaimed stream IDs?
+		}
+
+		if header.Status != protocol.OkSoFar {
+			client.mux.Unclaim(header.StreamID)
+		}
+
+		select {
+		case <-ctx.Done():
+			// TODO: wait for active requests to be completed and handle errors from client.conn.Close().
```

Comment:
First of all, thank you for the review!
It really helps to sort things out and I highly appreciate it.

I like the idea of having `Close` function, but I see a problem in proposed approach with cancellation logic inside `Cancel`:
Since `context` is passed to the client, it is expected that cancellation/deadline of that context will affect the client.
So, this workflow seems legit:
```go
ctx, cancel := context.WithCancel(ctx)
defer cancel()
client, err := client.NewClient(ctx,  "address")
// some logic and client usage
// ...
// cancel invokation
```

But thus the 
```go
    if err := cli.conn.Close(); err != nil { ... }
    if err := cli.mux.Close(); err != nil { ... }
```
lines inside `Cancel` will never be called.

I'm kinda new to Go and I'm not familiar with all of the conventions, and yes, it is highly likely that if there is `Cancel` user of the client will call it, especially if it's used in examples/documentation... But it still looks a bit confusing IMHO.

Maybe keep cancellation logic in `consume` and in `Cancel` simply call `context`'s `close`? But then it requires another goroutine or smth else to capture errors in `consume` and return them from `Cancel`.

Reply:
I am not claiming this is a completely 100% correct solution:
- https://play.golang.org/p/YTwxZRYYnzO

but it's the way I was imagining it :)
one can play with the condition at line 36 to enable an early-cancel by the user.

it seems to work (even under the `-race` detector :P)

you're right that, also in my example above, one needs some other medium (+possibly a goroutine) to communicate potential errors from the `consume` goroutine back to the `Close()` method.
perhaps a simple:
```go
type Client struct {
    // ... other fields elided ...
   mu sync.Mutex
   err error
}

func (cli *Client) consume() {
    for {
        select {
        default:
            if err := cli.conn.Close(); err != nil {
                cli.mu.Lock()
                if cli.err == nil { cli.err = err }
                cli.mu.Unlock()
            }
        }
    }
}
```

and return this `cli.err` error (guarded by the same mutex) in `Client.Close()`.

Reply:
Line 144 (`log.Printf("existing loop...")`) is *almost* never reached in your suggestion as far as I understand. :-/
Loop blocks on `io.ReadFull` and then returns due to:
```go
if ctx.Err() != nil {
    // something happened to the context.
    // ignore this error.
    return
}
```

I have changed `return` to `continue`, so loop successfully continues to the `case <-ctx.Done()` part.
However, it's still impossible to have all of the connection closing logic here because `io.ReadFull` will not unblock until the connection is not closed. 
Probably, it's possible to extract reading to another goroutine and `select` with `case <-ctx.Done()` and `case <- serverResponse~, but it looks a bit overcomplicated.

I see another 2 possibilities: 
1) Use context just for connecting, like `DialContext` does:
```
// If the context expires before
// the connection is complete, an error is returned. Once successfully
// connected, any expiration of the context will not affect the
// connection.
```

2) Make another goroutine for calling `Cancel` when the context is cancelled:
```go
go func() {
    <-ctx.Done()
    client.Cancel()
}
```
and remove `select` from `consume`, leaving only `for`-loop which returns on context error:
```go
func (client *Client) consume(ctx context.Context) {
    for {
        // ...
       if ctx.Err() != nil {
	    // something happened to the context.
	    // ignore this error.
            return
	}
    }
}
```

Looks like awaiting active requests should be done in the same place where `client.conn.Close()` is called, because after that `any blocked Read or Write operations will be unblocked and return errors`.

Reply:
I don't think it's an issue if the `"exiting loop..."` case isn't selected, as long as the goroutine is exited.
your modification of issueing `continue` in case of the read-header request will indeed make sure do go through that `<-ctx.Done()` case (and then exit, thus releasing resources associated with the goroutine).

having all the "cleanup/closing" of resources in one place (_ie:_ in `func (*Client) Close() error`) (and not in the `<-ctx.Done()` case) is easier to reason about, I think.